<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Algorithms for Automatized Detection of Hook Effect-bearing Amplification Curves • PCRedux</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Algorithms for Automatized Detection of Hook Effect-bearing Amplification Curves">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">PCRedux</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.2.6.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/PCRedux.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/PCRedux_visdat_pcrfit.html">PCRedux Package - Visualization Data Structure</a>
    </li>
    <li>
      <a href="../articles/SI1.html">Algorithms for Automatized Detection of Hook Effect-bearing Amplification Curves</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Algorithms for Automatized Detection of Hook Effect-bearing Amplification Curves</h1>
                        <h4 class="author">The PCRedux package authors</h4>
            
            <h4 class="date">2018-09-11</h4>
      
      
      <div class="hidden name"><code>SI1.Rmd</code></div>

    </div>

    
    

<p>The PDF version of this document is available <a href="https://github.com/devSJR/PCRedux/raw/master/docs/articles/SI1.pdf">online</a>.</p>
<div id="abstract" class="section level1">
<h1 class="hasAnchor">
<a href="#abstract" class="anchor"></a>Abstract</h1>
<p>This is a supplemental document for the study <em>Algorithms for Automatized Detection of Hook Effect-bearing Amplification Curves</em>. Quantitative real-time PCR (qPCR) is a widely used method for gene expression analysis, forensics and medical diagnostics <span class="citation">(Dvinge and Bertone 2009, <span class="citation">Martins et al. (2015)</span>, <span class="citation">Sauer, Reinke, and Courts (2016)</span>)</span>.</p>
<p>Numerous algorithms have been developed to extract features from amplification curves such as the cylce of quantification and the amplification efficiency <span class="citation">(Ruijter et al. 2013)</span>. There is an agreement, that these algorithms need to be evaluated and benchmarked for their performance <span class="citation">(Kemperman and McCall 2017)</span>. But at an earlier level it is important to have a solid foundation for the data preprocessing <span class="citation">(A.-N. Spiess et al. 2015, <span class="citation">A.-N. Spiess et al. (2016)</span>, <span class="citation">Ronde et al. (2017)</span>)</span>. Digitalization of processes holds the promise that potential human mistakes can be spotted and that diagnostic processes can be automatized.</p>
<p>The aim of the study is to provide software tools and algorithms, which assist qPCR users during the analysis and quality management of their data. In particular, this study shows how it is possible to automatically detect hook effects (see <span class="citation">Barratt and Mackay (2002)</span>) or hook effect-like curvatures.</p>
</div>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>The functions and data presented in the paper are available from <a href="https://github.com/devSJR/PCRedux" class="uri">https://github.com/devSJR/PCRedux</a>. The data, including the RDML file, are part of the  package and are made available in the CSV or RDML format <span class="citation">(Rödiger et al. 2017)</span> for vendor independent analysis.</p>
<p>All analyses were implemented and conducted with the <strong>R</strong> statistical computing language <span class="citation">(R Core Team 2017, <span class="citation">Rödiger et al. (2015)</span>)</span> and dedicated integrated development environments such as <strong>RKWard</strong> <span class="citation">(Rödiger et al. 2012)</span>. Further documentation can be found in the help files of the <strong>R</strong> packages.</p>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<p>The <strong>hookreg()</strong> and <strong>hookregNL()</strong> functions are part of the  package for the <strong>R</strong> statistical computing language. Download from CRAN <a href="http://cran.r-project.org/" class="uri">http://cran.r-project.org/</a> the <strong>R</strong> version for the required operating system and install <strong>R</strong>. Then start <strong>R</strong> and type in the prompt:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Select your local mirror</span>
<span class="kw">install.packages</span>(<span class="st">"PCRedux"</span>)</code></pre></div>
<p>The  package should just install. If this fails make sure you have write access to the destination directory and follow the instructions of the <strong>R</strong> documentation:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># The following command points to the help for download and install of packages </span>
<span class="co"># from CRAN-like repositories or from local files.</span>
?<span class="kw">install.packages</span>()</code></pre></div>
<p>The package can be installed as the latest development version using the <a href="https://cran.r-project.org/package=devtools">devtools</a> R package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Install devtools, if you haven't already.</span>
<span class="kw">install.packages</span>(<span class="st">"devtools"</span>)

<span class="kw">library</span>(devtools)
<span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="st">"devSJR/PCRedux"</span>)</code></pre></div>
<p>It is recommended to use software with an integrated development environment such as  <span class="citation">(Rödiger et al. 2012)</span>. To work with RDML data it is recommend to use the  package (<span class="math inline">\(\geq\)</span><sub>v.</sub>0.9-9) by invoking the <strong>rdmlEdit()</strong> function (for details see <span class="citation">Rödiger et al. (2017)</span>) or the rdmlEdit GUI web server (section ). The RDML file <code>hookreg.rdml</code> contains the amplification curve data. However, other software package (e.g., <span class="citation">(Lefever et al. 2009, <span class="citation">Ruijter et al. (2015)</span>)</span>) can also be used to work with the RMDL data file format.</p>
</div>
<div id="results-for-the-analysis-of-the-hookreg-rdml-data-set-by-humanrater" class="section level1">
<h1 class="hasAnchor">
<a href="#results-for-the-analysis-of-the-hookreg-rdml-data-set-by-humanrater" class="anchor"></a>Results for the analysis of the <code>hookreg.rdml</code> data set by <strong>humanrater()</strong>
</h1>
<p>All calculations in the following sections were employed on the <code>hookreg.rdml</code> data set, which is part of the  package. The data were transfered to the <strong>R</strong> environment by the  package <span class="citation">(Rödiger et al. 2017)</span>. An overview of the used samples and the qPCR detection chemistries and the classification by two humans (“Hook effect-like Rater 1”, “Hook effect-like Rater 2”) is shown in Table .</p>
<p>Loading experiment: exp1 run: run1 All amplification curves were plotted according to their experiment conditions. They differed in the target molecules (e.g., , ) and the detection chemistries (e.g., EvaGreen, SybrGreen, hydrolysis probes). Figure  shows seven plots for the corresponding experiments. The amplification curves were not preprocessed to preserve the curvature. Selected amplification curves were noisy (e.g., Figure F), had overshots or undershot in the background phase (e.g., Figure E-G), a short hook phase (e.g., Figure D). Amplifcation curves of Figure A, D, F and F exhibited a clearly visible hook effect or a hook like effect.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">2</span>))

<span class="co"># Plot all data of the hookreg.rdml-file according to their type.</span>
<span class="co"># Synthetic template, detected with Syto-13</span>
<span class="kw">matplot</span>(data[, <span class="dv">1</span>], data[, <span class="dv">2</span><span class="op">:</span><span class="dv">13</span>], <span class="dt">type=</span><span class="st">"l"</span>, <span class="dt">lty=</span><span class="dv">1</span>, <span class="dt">lwd=</span><span class="dv">2</span>, <span class="dt">ylab=</span><span class="st">"RFU"</span>, <span class="dt">xlab=</span><span class="st">"Cycle"</span>)
<span class="kw">mtext</span>(<span class="st">"A"</span>, <span class="dt">cex =</span> <span class="fl">1.8</span>, <span class="dt">side =</span> <span class="dv">3</span>, <span class="dt">adj =</span> <span class="dv">0</span>, <span class="dt">font =</span> <span class="dv">2</span>)

<span class="co"># Human MLC-2v, detected with a hydrolysis probe.</span>
<span class="kw">matplot</span>(data[, <span class="dv">1</span>], data[, <span class="dv">14</span><span class="op">:</span><span class="dv">45</span>], <span class="dt">type=</span><span class="st">"l"</span>, <span class="dt">lty=</span><span class="dv">1</span>, <span class="dt">lwd=</span><span class="dv">2</span>, <span class="dt">ylab=</span><span class="st">"RFU"</span>, <span class="dt">xlab=</span><span class="st">"Cycle"</span>)
<span class="kw">mtext</span>(<span class="st">"B"</span>, <span class="dt">cex =</span> <span class="fl">1.8</span>, <span class="dt">side =</span> <span class="dv">3</span>, <span class="dt">adj =</span> <span class="dv">0</span>, <span class="dt">font =</span> <span class="dv">2</span>)

<span class="co"># S27a housekeeping gene, detected with SybrGreen I.</span>
<span class="kw">matplot</span>(data[, <span class="dv">1</span>], data[, <span class="dv">46</span><span class="op">:</span><span class="dv">69</span>], <span class="dt">type=</span><span class="st">"l"</span>, <span class="dt">lty=</span><span class="dv">1</span>, <span class="dt">lwd=</span><span class="dv">2</span>, <span class="dt">ylab=</span><span class="st">"RFU"</span>, <span class="dt">xlab=</span><span class="st">"Cycle"</span>)
<span class="kw">mtext</span>(<span class="st">"C"</span>, <span class="dt">cex =</span> <span class="fl">1.8</span>, <span class="dt">side =</span> <span class="dv">3</span>, <span class="dt">adj =</span> <span class="dv">0</span>, <span class="dt">font =</span> <span class="dv">2</span>)

<span class="co"># Whole genome amplification, detected with EvaGreen.</span>
<span class="kw">matplot</span>(data[, <span class="dv">1</span>], data[, <span class="dv">70</span><span class="op">:</span><span class="dv">71</span>], <span class="dt">type=</span><span class="st">"l"</span>, <span class="dt">lty=</span><span class="dv">1</span>, <span class="dt">lwd=</span><span class="dv">2</span>, <span class="dt">ylab=</span><span class="st">"RFU"</span>, <span class="dt">xlab=</span><span class="st">"Cycle"</span>)
<span class="kw">mtext</span>(<span class="st">"D"</span>, <span class="dt">cex =</span> <span class="fl">1.8</span>, <span class="dt">side =</span> <span class="dv">3</span>, <span class="dt">adj =</span> <span class="dv">0</span>, <span class="dt">font =</span> <span class="dv">2</span>)

<span class="co"># Human BRCA1 gene, detected with a hydrolysis probe.</span>
<span class="kw">matplot</span>(data[, <span class="dv">1</span>], data[, <span class="dv">72</span><span class="op">:</span><span class="dv">87</span>], <span class="dt">type=</span><span class="st">"l"</span>, <span class="dt">lty=</span><span class="dv">1</span>, <span class="dt">lwd=</span><span class="dv">2</span>, <span class="dt">ylab=</span><span class="st">"RFU"</span>, <span class="dt">xlab=</span><span class="st">"Cycle"</span>)
<span class="kw">mtext</span>(<span class="st">"E"</span>, <span class="dt">cex =</span> <span class="fl">1.8</span>, <span class="dt">side =</span> <span class="dv">3</span>, <span class="dt">adj =</span> <span class="dv">0</span>, <span class="dt">font =</span> <span class="dv">2</span>)

<span class="co"># Human NRAS gene, detected with a hydrolysis probe.</span>
<span class="kw">matplot</span>(data[, <span class="dv">1</span>], data[, <span class="dv">88</span><span class="op">:</span><span class="dv">95</span>], <span class="dt">type=</span><span class="st">"l"</span>, <span class="dt">lty=</span><span class="dv">1</span>, <span class="dt">lwd=</span><span class="dv">2</span>, <span class="dt">ylab=</span><span class="st">"RFU"</span>, <span class="dt">xlab=</span><span class="st">"Cycle"</span>)
<span class="kw">mtext</span>(<span class="st">"F"</span>, <span class="dt">cex =</span> <span class="fl">1.8</span>, <span class="dt">side =</span> <span class="dv">3</span>, <span class="dt">adj =</span> <span class="dv">0</span>, <span class="dt">font =</span> <span class="dv">2</span>)

<span class="co"># Water control, detected with a hydrolysis probe.</span>
<span class="kw">matplot</span>(data[, <span class="dv">1</span>], data[, <span class="dv">96</span><span class="op">:</span><span class="dv">97</span>], <span class="dt">type=</span><span class="st">"l"</span>, <span class="dt">lty=</span><span class="dv">1</span>, <span class="dt">lwd=</span><span class="dv">2</span>, <span class="dt">ylab=</span><span class="st">"RFU"</span>, <span class="dt">xlab=</span><span class="st">"Cycle"</span>)
<span class="kw">mtext</span>(<span class="st">"G"</span>, <span class="dt">cex =</span> <span class="fl">1.8</span>, <span class="dt">side =</span> <span class="dv">3</span>, <span class="dt">adj =</span> <span class="dv">0</span>, <span class="dt">font =</span> <span class="dv">2</span>)</code></pre></div>
<div class="figure">
<img src="SI1_files/figure-html/ampcurves-1.png" alt="Amplification curves. A) Synthetic template, detected with Syto-13. B) Human \textit{MLC-2v}, detected with a hydrolysis probe. C) \textit{S27a} housekeeping gene, detected with SybrGreen I. D) Whole genome amplification, detected with EvaGreen. E) Human \textit{BRCA1} gene, detected with a hydrolysis probe. F) Human \textit{NRAS} gene, detected with a hydrolysis probe. G) Water control, detected with a hydrolysis probe. See Table \ref{Table_human_rated} for details. RFU, relative fluorescence units. \label{ampcurves}" width="696"><p class="caption">
Amplification curves. A) Synthetic template, detected with Syto-13. B) Human , detected with a hydrolysis probe. C)  housekeeping gene, detected with SybrGreen I. D) Whole genome amplification, detected with EvaGreen. E) Human  gene, detected with a hydrolysis probe. F) Human  gene, detected with a hydrolysis probe. G) Water control, detected with a hydrolysis probe. See Table  for details. RFU, relative fluorescence units. 
</p>
</div>
<p>Printout of all measured samples, their rating by two humans (rater 1 and rater 2) with their dichotomous ratings (0, no hook; 1, hook) and their sources.</p>
<ul>
<li>The boggy data (qpcR::boggy) set was taken from the  package <span class="citation">(Ritz and Spiess 2008, <span class="citation">A.-N. Spiess, Feig, and Ritz (2008)</span>)</span>.</li>
<li>The C127EGHP data (chipPCR::C127EGHP) set was taken from the  package <span class="citation">(Rödiger, Burdukiewicz, and Schierack 2015)</span>.</li>
<li>The testdat data (qpcR::testdat) set was taken from the  package <span class="citation">(Ritz and Spiess 2008, <span class="citation">A.-N. Spiess, Feig, and Ritz (2008)</span>)</span>.</li>
<li>Other data were prepared by Evrogen laboratory experiments.</li>
</ul>
</div>
<div id="results-for-the-analysis-with-hookreg-and-hookregnl" class="section level1">
<h1 class="hasAnchor">
<a href="#results-for-the-analysis-with-hookreg-and-hookregnl" class="anchor"></a>Results for the analysis with <strong>hookreg()</strong> and <strong>hookregNL()</strong>
</h1>
<p>This section contains the results of the analysis of the amplification curve data with the <strong>hookreg()</strong> function and the <strong>hookregNL()</strong> function. As in the previous sections, all code was commented to make it reproducible. Some rows in Table  and Table  appear to be empty. This expected behaviour may occur in cases where the corresponding functions were not able to calculate the coefficients due to a failed model fit or violation of the truncation criterion.</p>
<div id="results-for-the-analysis-of-the-hookreg-rdml-data-set-with-hookreg" class="section level2">
<h2 class="hasAnchor">
<a href="#results-for-the-analysis-of-the-hookreg-rdml-data-set-with-hookreg" class="anchor"></a>Results for the analysis of the <code>hookreg.rdml</code> data set with <strong>hookreg()</strong> </h2>
<p>The following code was used to analyze the <code>hookreg.rdml</code> data set with <strong>hookreg()</strong> function. The <strong>hookreg()</strong> function fits a linear model to a region of interest. The linear model is used to decide if the amplification curve as a hook effect or hook effect-like curvature.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Load PCRedux package to obtain the data and make the hookreg() function</span>
<span class="co"># available.</span>
<span class="kw">library</span>(PCRedux)
<span class="co"># Load the magrittr to use the %&gt;% pipe-operator</span>
<span class="kw">library</span>(magrittr)

<span class="co"># `data` is a temporary data frame of the hook.rdml amplification curve data file.</span>
<span class="co"># Apply the hookreg() function over the amplification curves and arrange the </span>
<span class="co"># results in the data frame `res_hookreg`.</span>

res_hookreg &lt;-<span class="st"> </span><span class="kw">sapply</span>(2L<span class="op">:</span><span class="kw">ncol</span>(data), <span class="cf">function</span>(i) {
    <span class="kw"><a href="../reference/hookreg.html">hookreg</a></span>(<span class="dt">x=</span>data[, <span class="dv">1</span>], <span class="dt">y=</span>data[, i])
}) <span class="op">%&gt;%</span><span class="st"> </span>t <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">sample=</span><span class="kw">colnames</span>(data)[<span class="op">-</span><span class="dv">1</span>],.)

<span class="co"># Fetch the calculated parameters from the calculations with the hookreg() </span>
<span class="co"># function as a table `res_hookreg_table`.</span>

res_hookreg_table &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">sample=</span><span class="kw">as.character</span>(res_hookreg[[<span class="st">"sample"</span>]]),
                                <span class="dt">intercept=</span><span class="kw">signif</span>(res_hookreg[[<span class="st">"intercept"</span>]], <span class="dv">2</span>),
                                <span class="dt">slope=</span><span class="kw">signif</span>(res_hookreg[[<span class="st">"slope"</span>]], <span class="dv">1</span>),
                                <span class="dt">hook.start=</span><span class="kw">signif</span>(res_hookreg[[<span class="st">"hook.start"</span>]], <span class="dv">0</span>),
                                <span class="dt">hook.delta=</span><span class="kw">signif</span>(res_hookreg[[<span class="st">"hook.delta"</span>]], <span class="dv">0</span>),
                                <span class="dt">p.value=</span><span class="kw">signif</span>(res_hookreg[[<span class="st">"p.value"</span>]], <span class="dv">4</span>),
                                <span class="dt">CI.low=</span><span class="kw">signif</span>(res_hookreg[[<span class="st">"CI.low"</span>]], <span class="dv">2</span>),
                                <span class="dt">CI.up=</span><span class="kw">signif</span>(res_hookreg[[<span class="st">"CI.up"</span>]], <span class="dv">2</span>),
                                <span class="dt">hook.fit=</span>res_hookreg[[<span class="st">"hook.fit"</span>]],
                                <span class="dt">hook.CI=</span>res_hookreg[[<span class="st">"hook.CI"</span>]],
                                <span class="dt">hook=</span>res_hookreg[[<span class="st">"hook"</span>]]
)</code></pre></div>
<p>Finally a pretty printout (Table ) of the results from the <strong>hookreg()</strong> function for the <code>hookreg.rdml</code> data set with the following code was prepared.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Load the xtable to create a LaTeX table from the `res_hookreg_table`.</span>
<span class="kw">library</span>(xtable)
<span class="kw">options</span>(<span class="dt">xtable.comment=</span><span class="ot">FALSE</span>)
<span class="kw">print</span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/xtable/topics/xtable">xtable</a></span>(res_hookreg_table, 
             <span class="dt">caption =</span> <span class="st">"Results from the hookreg() function for the hookreg.rdml </span>
<span class="st">             data set."</span>, 
             <span class="dt">label=</span><span class="st">'res_hookreg_table'</span>),
      <span class="dt">size =</span> <span class="st">"</span><span class="ch">\\</span><span class="st">tiny"</span>,
      <span class="dt">include.rownames =</span> <span class="ot">FALSE</span>,
      <span class="dt">include.colnames =</span> <span class="ot">TRUE</span>,
      <span class="dt">caption.placement =</span> <span class="st">"top"</span>,
      <span class="dt">comment=</span><span class="ot">FALSE</span>,
      <span class="dt">table.placement =</span> <span class="st">"!ht"</span>, <span class="dt">scalebox=</span><span class="st">'0.65'</span>
)</code></pre></div>

<p>The results of the <strong>hookreg()</strong> function are fairly comprehensive. The meaning of the columns is as followed:</p>

</div>
<div id="results-for-the-analysis-of-the-hookreg-rdml-data-set-with-hookregnl" class="section level2">
<h2 class="hasAnchor">
<a href="#results-for-the-analysis-of-the-hookreg-rdml-data-set-with-hookregnl" class="anchor"></a>Results for the analysis of the <code>hookreg.rdml</code> data set with <strong>hookregNL()</strong> </h2>
<p>The following code was used to analyze the <code>hookreg.rdml</code> data set with <strong>hookregNL()</strong> function. The procedure is similar to the analysis with the <strong>hookreg()</strong> function.</p>
<p>The <strong>hookreg()</strong> function fits a six parameter sigmoidal model to amplification curve. The non-linear model</p>
<p><span class="math display">\[
    f(x) = c + k \cdot x + \frac{d - c}{(1 + exp(b(log(x) - log(e))))^f}
    \newline
\]</span></p>
<p>is used to decide, based on the <em>k</em> parameter, if the amplification curve as a hook effect or hook effect-like curvature.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Note that the PCRedux package and the magrittr package need to be loaded (see above).</span>
<span class="co"># Load the qpcR package to prevent messages during the start.</span>
<span class="kw">suppressMessages</span>(<span class="kw">library</span>(qpcR))

<span class="co"># `data` is a temporary data frame of the hook.rdml amplification curve data file.</span>
<span class="co"># Apply the hookregNL() function over the amplification curves and arrange the </span>
<span class="co"># results in the data frame `res_hookregNL`.</span>
<span class="co"># Not that `suppressMessages()` to prevent warning messages from the qpcR package.</span>

res_hookregNL &lt;-<span class="st"> </span><span class="kw">suppressMessages</span>(<span class="kw">sapply</span>(2L<span class="op">:</span><span class="kw">ncol</span>(data), <span class="cf">function</span>(i) {
    <span class="kw"><a href="../reference/hookregNL.html">hookregNL</a></span>(<span class="dt">x=</span>data[, <span class="dv">1</span>], <span class="dt">y=</span>data[, i])
}) <span class="op">%&gt;%</span><span class="st"> </span>t <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">sample=</span><span class="kw">colnames</span>(data)[<span class="op">-</span><span class="dv">1</span>],.))

res_hookregNL_table &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">sample=</span><span class="kw">as.character</span>(res_hookregNL[[<span class="st">"sample"</span>]]),
                                  <span class="dt">slope=</span><span class="kw">signif</span>(<span class="kw">as.numeric</span>(res_hookregNL[[<span class="st">"slope"</span>]]), <span class="dv">1</span>),
                                  <span class="dt">CI.low=</span><span class="kw">signif</span>(<span class="kw">as.numeric</span>(res_hookregNL[[<span class="st">"CI.low"</span>]]), <span class="dv">2</span>),
                                  <span class="dt">CI.up=</span><span class="kw">signif</span>(<span class="kw">as.numeric</span>(res_hookregNL[[<span class="st">"CI.up"</span>]]), <span class="dv">2</span>),
                                  <span class="dt">hook.CI=</span><span class="kw">unlist</span>(res_hookregNL[[<span class="st">"hook"</span>]])
)</code></pre></div>
<p>Finally we prepare a pretty printout (Table ) of the results from the <strong>hookregNL()</strong> function for the <code>hookreg.rdml</code> data set with the following code with the code shown next.</p>
<p>The results of the <strong>hookregNL()</strong> function are less comprehensive then from the <strong>hookreg()</strong> function . The meaning of the columns is as followed:</p>

<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(xtable)
<span class="kw">options</span>(<span class="dt">xtable.comment=</span><span class="ot">FALSE</span>)

<span class="kw">print</span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/xtable/topics/xtable">xtable</a></span>(res_hookregNL_table, 
             <span class="dt">caption =</span> <span class="st">"Results from the hookregNL() function for the </span>
<span class="st">             hookreg.rdml data set."</span>, 
             <span class="dt">label=</span><span class="st">'res_hookregNL_table'</span>),
      <span class="dt">size =</span> <span class="st">"</span><span class="ch">\\</span><span class="st">tiny"</span>,
      <span class="dt">include.rownames =</span> <span class="ot">FALSE</span>,
      <span class="dt">include.colnames =</span> <span class="ot">TRUE</span>,
      <span class="dt">caption.placement =</span> <span class="st">"top"</span>,
      <span class="dt">comment=</span><span class="ot">FALSE</span>,
      <span class="dt">table.placement =</span> <span class="st">"!ht"</span>, <span class="dt">scalebox=</span><span class="st">'0.65'</span>
)</code></pre></div>

</div>
</div>
<div id="comparison-of-the-hookreg-and-hookregnl-methods" class="section level1">
<h1 class="hasAnchor">
<a href="#comparison-of-the-hookreg-and-hookregnl-methods" class="anchor"></a>Comparison of the <strong>hookreg()</strong> and <strong>hookregNL()</strong> methods</h1>
<p>The decisions from the human classification (see Table ) and the results from the machine decision (section  and section ) were aggregated in Table .</p>
<p>Finally a pretty printout (Table ) of the aggregated data set with the following code was prepared:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># A simple logic was applied to improve the classification result. In this case</span>
<span class="co"># the assumption was, that an amplification curve has an hook effect or hook effect-like</span>
<span class="co"># curvature, if either the hookreg() or hookregNL() function are positive.</span>

meta_hookreg &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(res), <span class="cf">function</span>(i){
    <span class="kw">ifelse</span>(res[i, <span class="st">"hookreg"</span>] <span class="op">==</span><span class="st"> </span><span class="dv">1</span> <span class="op">||</span><span class="st"> </span>res[i, <span class="st">"hookregNL"</span>] <span class="op">==</span><span class="st"> </span><span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">0</span>)
})

res_out &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">Sample=</span>res[[<span class="st">"Sample"</span>]], res[[<span class="st">"Human rater"</span>]], 
                      res_hookreg[[<span class="st">"hook"</span>]], res_hookregNL_table[[<span class="st">"hook.CI"</span>]], 
                      meta_hookreg)

<span class="kw">colnames</span>(res_out) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"Sample"</span>,
                       <span class="st">"Human rater"</span>,
                       <span class="st">"hookreg"</span>,
                       <span class="st">"hookregNL"</span>,
                       <span class="st">"hookreg and hoohkreNL combined"</span>
)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(xtable)
<span class="kw">options</span>(<span class="dt">xtable.comment=</span><span class="ot">FALSE</span>)

<span class="kw">print</span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/xtable/topics/xtable">xtable</a></span>(res_out, <span class="dt">digits=</span><span class="dv">0</span>, 
             <span class="dt">caption =</span> <span class="st">"Aggregated decisions from the human classification and </span>
<span class="st">the results from the machine decision of the hookreg() and hookregNL() </span>
<span class="st">functions."</span>, <span class="dt">label=</span><span class="st">'method_comparision'</span>), ,
      <span class="dt">caption.placement =</span> <span class="st">"top"</span>,
      <span class="dt">scalebox=</span><span class="st">'0.65'</span>)</code></pre></div>

<p>The performance of the <strong>hookreg()</strong> and <strong>hookregNL()</strong> functions was analyzed with the <strong>performeR()</strong> function of the  package (Table ). The methods were adopted from <span class="citation">Brenner and Gefeller (1997)</span> and <span class="citation">Kuhn (2008)</span>. Note that the formula for the calculations of the sensitivity, specificity, precision, Negative predictive value, fall-out, alse negative rate, false discovery rate, Accuracy, F1 score, Matthews correlation coefficient and kappa by Cohen are described in the documentation of the  package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">res_performeR &lt;-<span class="st"> </span><span class="kw">rbind</span>(
    <span class="dt">hookreg=</span><span class="kw"><a href="../reference/performeR.html">performeR</a></span>(res_out[[<span class="st">"hookreg"</span>]], res_out[[<span class="st">"Human rater"</span>]]),
                       <span class="dt">hookregNL=</span><span class="kw"><a href="../reference/performeR.html">performeR</a></span>(res_out[[<span class="st">"hookregNL"</span>]], res_out[[<span class="st">"Human rater"</span>]]),
                       <span class="dt">combined_hookreg=</span><span class="kw"><a href="../reference/performeR.html">performeR</a></span>(res_out[[<span class="st">"hookreg and hoohkreNL combined"</span>]], 
                                                  res_out[[<span class="st">"Human rater"</span>]])
) <span class="op">%&gt;%</span><span class="st"> </span>t <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">signif</span>(<span class="dv">4</span>)

<span class="kw">colnames</span>(res_performeR) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"hookreg"</span>, <span class="st">"hookregNL"</span>, <span class="st">"hookreg and hookregNL"</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(xtable)
<span class="kw">options</span>(<span class="dt">xtable.comment=</span><span class="ot">FALSE</span>)

<span class="kw">print</span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/xtable/topics/xtable">xtable</a></span>(res_performeR, <span class="dt">digits=</span><span class="dv">4</span>, 
             <span class="dt">caption =</span> <span class="st">"Analysis of the performance of both algorithms. The </span>
<span class="st">performance of the individual test and the combination of the tests is shown. </span>
<span class="st">Note that the classification improved if the hookreg() and hookregNL() function </span>
<span class="st">were combined by a logical statement. The measure were determined with the </span>
<span class="ch">\\</span><span class="st">textit{performeR()} function from the </span><span class="ch">\\</span><span class="st">texttt{PCRedux} package. Sensitivity, </span>
<span class="st">TPR; Specificity, SPC; Precision, PPV; Negative predictive value, NPV; Fall-out, </span>
<span class="st">FPR; False negative rate, FNR; False discovery rate, FDR; Accuracy, ACC; F1 </span>
<span class="st">score, F1; Matthews correlation coefficient, MCC, Cohen's kappa (binary </span>
<span class="st">classification), $</span><span class="ch">\\</span><span class="st">kappa$"</span>, <span class="dt">label=</span><span class="st">'res_performeR'</span>),
      <span class="dt">size =</span> <span class="st">"normalsize"</span>,
      <span class="dt">include.rownames =</span> <span class="ot">TRUE</span>,
      <span class="dt">include.colnames =</span> <span class="ot">TRUE</span>,
      <span class="dt">caption.placement =</span> <span class="st">"top"</span>,
      <span class="dt">comment=</span><span class="ot">FALSE</span>,
      <span class="dt">table.placement =</span> <span class="st">"!ht"</span>, <span class="dt">scalebox=</span><span class="st">'0.75'</span>
)</code></pre></div>

</div>
<div id="funding" class="section level1">
<h1 class="hasAnchor">
<a href="#funding" class="anchor"></a>Funding</h1>
<p>This work was funded by the Federal Ministry of Education and Research (BMBF) InnoProfile-Transfer-Project 03IPT611X and in part by “digilog: Digitale und analoge Begleiter für eine alternde Bevölkerung” (Gesundheitscampus Brandenburg, Brandenburg Ministry for Science, Research and Culture). We thank Franziska Dinter (BTU) for reevaluation of the amplification curve data and Maria Tokarenko (Evrogen JSC) for wet lab experiments conduction.</p>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-barratt_improving_2002">
<p>Barratt, Kevin, and John F. Mackay. 2002. “Improving Real-Time PCR Genotyping Assays by Asymmetric Amplification.” <em>Journal of Clinical Microbiology</em> 40 (4): 1571–2. doi:<a href="https://doi.org/10.1128/JCM.40.4.1571-1572.2002">10.1128/JCM.40.4.1571-1572.2002</a>.</p>
</div>
<div id="ref-brenner_variation_1997">
<p>Brenner, Hermann, and Olaf Gefeller. 1997. “Variation of sensitivity, specificity, likelihood ratios and predictive values with disease prevalence.” <em>Statistics in Medicine</em> 16 (9): 981–91. <a href="http://www.floppybunny.org/robin/web/virtualclassroom/stats/basics/articles/odds_risks/odds_sensitivity)likelihood_ratios_validity_brenner_1997.pdf" class="uri">http://www.floppybunny.org/robin/web/virtualclassroom/stats/basics/articles/odds_risks/odds_sensitivity)likelihood_ratios_validity_brenner_1997.pdf</a>.</p>
</div>
<div id="ref-dvinge_htqpcr:_2009">
<p>Dvinge, Heidi, and Paul Bertone. 2009. “HTqPCR: high-throughput analysis and visualization of quantitative real-time PCR data in R.” <em>Bioinformatics</em> 25 (24): 3325–6. doi:<a href="https://doi.org/10.1093/bioinformatics/btp578">10.1093/bioinformatics/btp578</a>.</p>
</div>
<div id="ref-kemperman_mircomp-shiny:_2017">
<p>Kemperman, Lauren, and Matthew N. McCall. 2017. “miRcomp-Shiny: Interactive assessment of qPCR-based microRNA quantification and quality control algorithms.” <em>F1000Research</em> 6 (November): 2046. <a href="https://f1000research.com/articles/6-2046/v1" class="uri">https://f1000research.com/articles/6-2046/v1</a>.</p>
</div>
<div id="ref-kuhn_building_2008">
<p>Kuhn, Max. 2008. “Building Predictive Models in R Using the caret Package.” <em>Journal of Statistical Software</em> 28 (5). <a href="http://www.jstatsoft.org/v28/i05/" class="uri">http://www.jstatsoft.org/v28/i05/</a>.</p>
</div>
<div id="ref-lefever_rdml_2009">
<p>Lefever, Steve, Jan Hellemans, Filip Pattyn, Daniel R. Przybylski, Chris Taylor, René Geurts, Andreas Untergasser, Jo Vandesompele, and RDML on behalf of the Consortium. 2009. “RDML: structured language and reporting guidelines for real-time quantitative PCR data.” <em>Nucleic Acids Research</em> 37 (7): 2065–9. doi:<a href="https://doi.org/10.1093/nar/gkp056">10.1093/nar/gkp056</a>.</p>
</div>
<div id="ref-martins_dna_2015">
<p>Martins, C., G. Lima, Mr. Carvalho, L. Cainé, and Mj. Porto. 2015. “DNA quantification by real-time PCR in different forensic samples.” <em>Forensic Science International: Genetics Supplement Series</em> 5 (December): e545–e546. doi:<a href="https://doi.org/10.1016/j.fsigss.2015.09.215">10.1016/j.fsigss.2015.09.215</a>.</p>
</div>
<div id="ref-R_language">
<p>R Core Team. 2017. <em>R: A Language and Environment for Statistical Computing</em>. Vienna, Austria: R Foundation for Statistical Computing. <a href="https://www.R-project.org/" class="uri">https://www.R-project.org/</a>.</p>
</div>
<div id="ref-Ritz2008">
<p>Ritz, Christian, and Andrej-Nikolai Spiess. 2008. “qpcR: an R package for sigmoidal model selection in quantitative real-time polymerase chain reaction analysis.” <em>Bioinformatics</em> 24 (13): 1549–51. doi:<a href="https://doi.org/10.1093/bioinformatics/btn227">10.1093/bioinformatics/btn227</a>.</p>
</div>
<div id="ref-ronde_practical_2017">
<p>Ronde, Maurice W. J. de, Jan M. Ruijter, David Lanfear, Antoni Bayes-Genis, Maayke G. M. Kok, Esther E. Creemers, Yigal M. Pinto, and Sara-Joan Pinto-Sietsma. 2017. “Practical data handling pipeline improves performance of qPCR-based circulating miRNA measurements.” <em>RNA</em> 23 (5): 811–21. doi:<a href="https://doi.org/10.1261/rna.059063.116">10.1261/rna.059063.116</a>.</p>
</div>
<div id="ref-roediger2015chippcr">
<p>Rödiger, Stefan, Michał Burdukiewicz, and Peter Schierack. 2015. “chipPCR: an R package to pre-process raw data of amplification curves.” <em>Bioinformatics</em> 31 (17): 2900–2902. doi:<a href="https://doi.org/10.1093/bioinformatics/btv205">10.1093/bioinformatics/btv205</a>.</p>
</div>
<div id="ref-roediger2015r">
<p>Rödiger, Stefan, Michał Burdukiewicz, Konstantin A. Blagodatskikh, and Peter Schierack. 2015. “R as an Environment for the Reproducible Analysis of DNA Amplification Experiments.” <em>The R Journal</em> 7 (2): 127–50. <a href="http://journal.r-project.org/archive/2015-1/RJ-2015-1.pdf" class="uri">http://journal.r-project.org/archive/2015-1/RJ-2015-1.pdf</a>.</p>
</div>
<div id="ref-roediger_enabling_2017">
<p>Rödiger, Stefan, Michał Burdukiewicz, Andrej-Nikolai Spiess, and Konstantin Blagodatskikh. 2017. “Enabling reproducible real-time quantitative PCR research: the RDML package.” <em>Bioinformatics</em>, August. doi:<a href="https://doi.org/10.1093/bioinformatics/btx528">10.1093/bioinformatics/btx528</a>.</p>
</div>
<div id="ref-roediger_rkward_2012">
<p>Rödiger, Stefan, Thomas Friedrichsmeier, Prasenjit Kapat, and Meik Michalke. 2012. “RKWard: a comprehensive graphical user interface and integrated development environment for statistical analysis with R.” <em>Journal of Statistical Software</em> 49 (9): 1–34. <a href="https://www.jstatsoft.org/article/view/v049i09/v49i09.pdf" class="uri">https://www.jstatsoft.org/article/view/v049i09/v49i09.pdf</a>.</p>
</div>
<div id="ref-ruijter_rdml-ninja_2015">
<p>Ruijter, Jan M., Steve Lefever, Jasper Anckaert, Jan Hellemans, Michael W. Pfaffl, Vladimir Benes, Stephen A. Bustin, Jo Vandesompele, Andreas Untergasser, and RDML on behalf of the Consortium. 2015. “RDML-Ninja and RDMLdb for standardized exchange of qPCR data.” <em>BMC Bioinformatics</em> 16 (1): 197. doi:<a href="https://doi.org/10.1186/s12859-015-0637-6">10.1186/s12859-015-0637-6</a>.</p>
</div>
<div id="ref-ruijter_evaluation_2013">
<p>Ruijter, Jan M., Michael W. Pfaffl, Sheng Zhao, Andrej N. Spiess, Gregory Boggy, Jochen Blom, Robert G. Rutledge, et al. 2013. “Evaluation of qPCR curve analysis methods for reliable biomarker discovery: Bias, resolution, precision, and implications.” <em>Methods</em> 59 (1): 32–46. doi:<a href="https://doi.org/10.1016/j.ymeth.2012.08.011">10.1016/j.ymeth.2012.08.011</a>.</p>
</div>
<div id="ref-sauer_differentiation_2016">
<p>Sauer, Eva, Ann-Kathrin Reinke, and Cornelius Courts. 2016. “Differentiation of five body fluids from forensic samples by expression analysis of four microRNAs using quantitative PCR.” <em>Forensic Science International: Genetics</em> 22 (May): 89–99. doi:<a href="https://doi.org/10.1016/j.fsigen.2016.01.018">10.1016/j.fsigen.2016.01.018</a>.</p>
</div>
<div id="ref-spiess_impact_2015">
<p>Spiess, Andrej-Nikolai, Claudia Deutschmann, Michał Burdukiewicz, Ralf Himmelreich, Katharina Klat, Peter Schierack, and Stefan Rödiger. 2015. “Impact of Smoothing on Parameter Estimation in Quantitative DNA Amplification Experiments.” <em>Clinical Chemistry</em> 61 (2): 379–88. doi:<a href="https://doi.org/10.1373/clinchem.2014.230656">10.1373/clinchem.2014.230656</a>.</p>
</div>
<div id="ref-spiess_highly_2008">
<p>Spiess, Andrej-Nikolai, Caroline Feig, and Christian Ritz. 2008. “Highly accurate sigmoidal fitting of real-time PCR data by introducing a parameter for asymmetry.” <em>BMC Bioinformatics</em> 9 (1): 221. doi:<a href="https://doi.org/10.1186/1471-2105-9-221">10.1186/1471-2105-9-221</a>.</p>
</div>
<div id="ref-spiess_system-specific_2016">
<p>Spiess, Andrej-Nikolai, Stefan Rödiger, Michał Burdukiewicz, Thomas Volksdorf, and Joel Tellinghuisen. 2016. “System-specific periodicity in quantitative real-time polymerase chain reaction data questions threshold-based quantitation.” <em>Scientific Reports</em> 6 (December): 38951. doi:<a href="https://doi.org/10.1038/srep38951">10.1038/srep38951</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#abstract">Abstract</a></li>
      <li><a href="#introduction">Introduction</a></li>
      <li><a href="#installation">Installation</a></li>
      <li><a href="#results-for-the-analysis-of-the-hookreg-rdml-data-set-by-humanrater">Results for the analysis of the <code>hookreg.rdml</code> data set by <strong>humanrater()</strong></a></li>
      <li>
<a href="#results-for-the-analysis-with-hookreg-and-hookregnl">Results for the analysis with <strong>hookreg()</strong> and <strong>hookregNL()</strong></a><ul class="nav nav-pills nav-stacked">
<li><a href="#results-for-the-analysis-of-the-hookreg-rdml-data-set-with-hookreg">Results for the analysis of the <code>hookreg.rdml</code> data set with <strong>hookreg()</strong> </a></li>
      <li><a href="#results-for-the-analysis-of-the-hookreg-rdml-data-set-with-hookregnl">Results for the analysis of the <code>hookreg.rdml</code> data set with <strong>hookregNL()</strong> </a></li>
      </ul>
</li>
      <li><a href="#comparison-of-the-hookreg-and-hookregnl-methods">Comparison of the <strong>hookreg()</strong> and <strong>hookregNL()</strong> methods</a></li>
      <li><a href="#funding">Funding</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Stefan Roediger, Michal Burdukiewicz, Andrej-Nikolai Spiess, Konstantin A. Blagodatskikh.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
